javalanglib/src/main/scala/java/lang/_String.scala

```diff
@@ -319,14 +322,302 @@ final class _String private () // scalastyle:ignore
+   * locale-insensitive. This is correct because those two operations are
+   * equivalent for English and for locale-insensitive. In fact, only
+   * Lithuanian (lt), Turkish (tr) and Azeri (az) have different behaviors than
+   * locale-insensitive.
+   *
+   * The overloads with a `Locale` specificially test for those three languages
+   * and delegate to dedicated methods to handle them. Those methods start by
+   * handling their respective special cases, then delegate to the locale-
+   * insensitive version. The special cases are specified in the Unicode
+   * reference file at
+   *
+   *   https://unicode.org/Public/13.0.0/ucd/SpecialCasing.txt
+   *
+   * That file first contains a bunch of locale-insensitive special cases,
+   * which we do not need to handle. Only the last two sections about locale-
+   * sensitive special-cases are important for us.
+   *
+   * Some of the rules are further context-sensitive, using predicates that are
+   * defined in Section 3.13 "Default Case Algorithms" of the Unicode Standard,
+   * available at
+   *
+   *   http://www.unicode.org/versions/Unicode13.0.0/
+   *
+   * We based the implementations on Unicode 13.0.0. It is worth noting that
+   * there has been no non-comment changes in the SpecialCasing.txt file
+   * between Unicode 4.1.0 and 13.0.0 (perhaps even earlier; the version 4.1.0
+   * is the earliest that is easily accessible).
+   */
+
+  def toLowerCase(locale: Locale): String = {
+    locale.getLanguage() match {
+      case "lt"        => toLowerCaseLithuanian()
+      case "tr" | "az" => toLowerCaseTurkishAndAzeri()
+      case _           => toLowerCase()
+    }
+  }
+
+  private def toLowerCaseLithuanian(): String = {
+    /* Relevant excerpt from SpecialCasing.txt
+     *
+     * # Lithuanian
+     *
+     * # Lithuanian retains the dot in a lowercase i when followed by accents.
+     *
+     * [...]
+     *
+     * # Introduce an explicit dot above when lowercasing capital I's and J's
+     * # whenever there are more accents above.
+     * # (of the accents used in Lithuanian: grave, acute, tilde above, and ogonek)
+     *
+     * 0049; 0069 0307; 0049; 0049; lt More_Above; # LATIN CAPITAL LETTER I
+     * 004A; 006A 0307; 004A; 004A; lt More_Above; # LATIN CAPITAL LETTER J
+     * 012E; 012F 0307; 012E; 012E; lt More_Above; # LATIN CAPITAL LETTER I WITH OGONEK
+     * 00CC; 0069 0307 0300; 00CC; 00CC; lt; # LATIN CAPITAL LETTER I WITH GRAVE
+     * 00CD; 0069 0307 0301; 00CD; 00CD; lt; # LATIN CAPITAL LETTER I WITH ACUTE
+     * 0128; 0069 0307 0303; 0128; 0128; lt; # LATIN CAPITAL LETTER I WITH TILDE
+     */
+
+    def moreAbove(i: Int): scala.Boolean = {
+      // scalastyle:off return
+      import Character._
+      val len = length()
+      var j = i + 1
+      while (j != len) {
+        val cp = codePointAt(j)
+        combiningClassNoneOrAboveOrOther(cp) match {
+          case CombiningClassIsNone  => return false
```

Comment:
Are you sure it is correct to early abort here? IIUC under the Canonical Ordering Algorithm, combining marks that have a combining class of zero are *not* reordered. As such, it seems they can appear in *any* location. (My understanding and assumption is that there *are* combining marks with combining class none).

Reply:
Yes, I'm sure. The algorithm here is independent of any other properties of Unicode, including the Canonical Ordering. The Contexts used in the Default Case Algorithms are entirely specified in Section 3.13 of the Standard, and they are very clear that a character with combining class 0 is not allowed in this context.

I have added comments to all the functions implementing Contexts with the definitions from Section 3.13.

I have also added tests that ensure that combining mark characters with combining class 0 do cut the link in the Contexts.